game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java

```diff
@@ -0,0 +1,143 @@
+package games.strategy.triplea.delegate.battle.steps.fire;
+
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.triplea.delegate.Matches;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import lombok.Value;
+import org.triplea.java.collections.CollectionUtils;
+
+/**
+ * Group of units that are firing on targets
+ *
+ * <p>If the group is suicideOnHit, then all of the units should have the same unit type
+ */
+@Value
+public class FiringGroup {
+  String displayName;
+  String groupName;
+  Collection<Unit> firingUnits;
+  Collection<Unit> targetUnits;
+  boolean suicideOnHit;
+
+  public Collection<Unit> getTargetUnits() {
+    return Collections.unmodifiableCollection(targetUnits);
+  }
+
+  public Collection<Unit> getFiringUnits() {
+    return Collections.unmodifiableCollection(firingUnits);
+  }
+
+  /**
+   * Keeps alive units around
+   *
+   * <p>Units unfortunately don't track their own status so the list of targets needs to be updated
+   * as the battle progresses.
+   *
+   * @param aliveUnits Units that are still alive
+   */
+  public void retainAliveTargets(final Collection<Unit> aliveUnits) {
+    targetUnits.retainAll(aliveUnits);
+  }
+
+  /**
+   * Splits up the firingUnits by suicideOnHit status and groups them by unit type
+   *
+   * @param name Name of the firing units
+   * @param firingUnits Collection of units that are firing
+   * @param targetUnits Collection of units that are being hit
+   * @return List of FiringGroup
+   */
+  public static List<FiringGroup> groupBySuicideOnHit(
+      final String name, final Collection<Unit> firingUnits, final Collection<Unit> targetUnits) {
+    final List<FiringGroup> groups = new ArrayList<>();
+    final List<Collection<Unit>> separatedBySuicide = separateSuicideOnHit(firingUnits);
+
+    // ensure each firing group has a unique name by adding prefixes
```

Comment:
Where is the prefix added? It's an interesting comment, though scanning down it's not immediately obvious where/what this is refererring to exactly.

Reply:
The prefixes are added inside of the callees `generateFiringGroupsWithOneSuicideAndOneNonSuicide` and `generateFiringGroups`.

Reply:
Cool, though that seems to be a detail of those methods. At this level, it's not obvious that is happening, therefore this comment seems out of place. One rule of thumb, any method should be understandable by simply reading it and not having to also read the methods that it invokes. One can see if that rule is violated then there is a scaling problem to understand a given method if you have to read all of the called methods, and all of the methods that are in turn called to the n'th degree.

Reply:
It was there to explain why the if/else if/else.  The only reason for them is to generate suffices.  If I didn't care about having unique suffices, I would just do a simple
```
return separatedBySuicide.stream()
        .map(newFiringUnits -> new FiringGroup(name, newFiringUnits, targetUnits))
        .collect(Collectors.toList());
```

But I've been thinking about your comments here and I think if I instead change the above to:
```
return separatedBySuicide.stream()
        .map(newFiringUnits -> new FiringGroup(generateName(name, newFiringUnits, separatedBySuicide), newFiringUnits, targetUnits))
        .collect(Collectors.toList());
```

Then the logic for the unique names is more obviously contained to just dealing with the string instead of generating firing groups.